src/Dfe.PlanTech.AzureFunctions/Functions/QueueReceiver.cs

```diff
@@ -11,162 +11,198 @@
-            {
-                await ProcessMessage(message, messageActions, cancellationToken);
-            }
+            await ProcessMessage(message, messageActions, cancellationToken);
         }
+    }
 
-        private async Task ProcessMessage(ServiceBusReceivedMessage message, ServiceBusMessageActions messageActions, CancellationToken cancellationToken)
+    private async Task ProcessMessage(ServiceBusReceivedMessage message, ServiceBusMessageActions messageActions, CancellationToken cancellationToken)
+    {
+        try
         {
-            try
-            {
-                string cmsEvent = GetCmsEvent(message.Subject);
-                var text = Encoding.UTF8.GetString(message.Body);
+            string cmsEvent = GetCmsEvent(message.Subject);
 
-                Logger.LogInformation("Performing Action: {action}", cmsEvent);
-                Logger.LogInformation("Processing {text}", text);
+            Logger.LogInformation("CMS Event: {cmsEvent}", cmsEvent);
 
-                ContentComponentDbEntity mapped = _mappers.ToEntity(text);
-                ContentComponentDbEntity? existing = await GetExistingDbEntity(mapped, cancellationToken);
-
-                if (existing != null)
-                {
-                    mapped.Archived = existing.Archived;
-                    mapped.Published = existing.Published;
-                    mapped.Deleted = existing.Deleted;
-                }
-
-                switch (cmsEvent)
-                {
-                    case "create":
-                    case "save":
-                    case "auto_save":
-                        break;
-                    case "archive":
-                        mapped.Archived = true;
-                        break;
-                    case "unarchive":
-                        mapped.Archived = false;
-                        break;
-                    case "publish":
-                        mapped.Published = true;
-                        break;
-                    case "unpublish":
-                        mapped.Published = false;
-                        break;
-                    case "delete":
-                        mapped.Deleted = true;
-                        break;
-                    default:
-                        throw new CmsEventException(string.Format("CMS Event \"{0}\" not implemented", cmsEvent));
-                }
+            ContentComponentDbEntity mapped = GetMapped(message);
+            ContentComponentDbEntity? existing = await GetExisting(mapped, cancellationToken);
 
-                long rowsChanged = await UpsertEntityInDatabase(mapped, existing, cancellationToken);
+            if (cmsEvent.Equals("unpublish") && existing != null)
+            {
+                existing.Published = false;
+            }
+            else
+            {
+                ProcessCmsEvent(cmsEvent, ref mapped);
 
-                if (rowsChanged == 0L)
+                if (existing == null)
```

Comment:
IMO pull this if statement out into own method as well (e.g. `UpsertEntity`). Then we the `ProcessMessage` method can look like:

```
            ContentComponentDbEntity mapped = GetMapped(message);
            ContentComponentDbEntity? existing = await GetExisting(mapped, cancellationToken);
             ProcessCmsEvent(cmsEvent, ref mapped); // ???? changed/tweaked
            UpsertEntity(mapped, existing);
            await DbSaveChanges(cancellationToken);
            await messageActions.CompleteMessageAsync(message, cancellationToken);
```

It keeps the logic for everything else to their own methods, and then just the high level process logic in the main method (i.e. map -> process cms event -> upsert -> save).

I mean in an ideal world half of these methods would be their own classes but I digress; no need to overengineer I guess!